tla-assignments/src/main/scala/at/forsyte/apalache/tla/assignments/EnabledRewriter.scala

```diff
@@ -0,0 +1,288 @@
+package at.forsyte.apalache.tla.assignments
+
+import at.forsyte.apalache.tla.lir._
+import at.forsyte.apalache.tla.imp.src.SourceStore
+import at.forsyte.apalache.tla.lir.storage.ChangeListener
+import at.forsyte.apalache.tla.lir.storage.SourceLocator
+import at.forsyte.apalache.tla.lir.storage.BodyMapFactory
+import at.forsyte.apalache.tla.assignments.SmtFreeSymbolicTransitionExtractor
+import at.forsyte.apalache.tla.lir.oper.TlaActionOper
+import at.forsyte.apalache.tla.lir.oper.ApalacheOper
+import at.forsyte.apalache.tla.lir.values.TlaBool
+import at.forsyte.apalache.tla.lir.oper.TlaBoolOper
+import at.forsyte.apalache.tla.pp.temporal.utils
+import at.forsyte.apalache.tla.lir.transformations.{TlaExTransformation, TransformationTracker}
+import com.typesafe.scalalogging.LazyLogging
+import at.forsyte.apalache.tla.pp.NotInKeraError
+import at.forsyte.apalache.tla.pp.ConstSimplifier
+import at.forsyte.apalache.tla.pp.Normalizer
+
+/**
+ * Attempts to rewrite `ENABLED foo` operators into formulas that are true when action `foo` is enabled.
+ *
+ * @author
+ *   Philip Offtermatt
+ */
+class EnabledRewriter(
+    tracker: TransformationTracker,
+    sourceStore: SourceStore,
+    changeListener: ChangeListener,
+    module: TlaModule)
+    extends TlaExTransformation with LazyLogging {
+
+  /**
+   * Removes the assignments x' := foo from an expression by replacing them with TRUE
+   */
+  private def removeAssignmentsFromExpression(ex: TlaEx): TlaEx = {
+    ex match {
+      case OperEx(oper, args @ _*) =>
+        oper match {
+          case ApalacheOper.assign =>
+            ValEx(TlaBool(true))(Typed(BoolT1))
+          case _ =>
+            OperEx(oper, args.map(removeAssignmentsFromExpression): _*)(ex.typeTag)
+        }
+      case _ =>
+        ex
+    }
+  }
+
+  /**
+   * Transforms an expression by moving all boolean quantification to the very outside.
+   * For example
+   *   {{{
+   *   (∃i$2 ∈ (1 .. 5): (x' = i$2)) ∧ (x' ≥ 0)
+   *   }}} becomes
+   * {{{
+   *   ∃i$2 ∈ (1 .. 5): ((x' = i$2) ∧ (x' ≥ 0))
+   *   }}}.
+   *
+   * The order is kept, in that afterwards, inner quantifiers are still inside:
+   *    {{{
+   *   (∃i$2 ∈ (1 .. 5): (x' = i$2) /\ ∃i$3 ∈ (1 .. 4): (y' = i$3)) ∧ (x' ≥ 0)
+   *   }}}
+   * becomes
+   *   {{{
+   *   ∃i$2 ∈ (1 .. 5): ∃i$3 ∈ (1 .. 4): (((x' = i$2) /\ (y' = i$3)) ∧ (x' ≥ 0))
+   *   }}}
+   */
+  private def putQuantificationsOutward(ex: TlaEx): TlaEx = {
+    def collectAndRemoveQuantifications(ex: TlaEx): (Seq[OperEx], TlaEx) = {
```

Comment:
I dislike using TlaBoolOper, since it gives (at least me) the feeling this would somehow relate to all boolean operators,
when it relates to very specific ones. (Using OperEx has a similar issue, but since its scope is bigger, I think an "artificial restriction" feels more natural).

I'll rather forget abound unbounded quantification and make an explicit note of that fact.